lib/explorer/data_frame.ex

```diff
@@ -1558,15 +1559,79 @@ defmodule Explorer.DataFrame do
         x integer [1, 2, 3, 4, 5, "..."]
         y string ["a", "b", "c", "d", "e", "..."]
       >
+
+      iex> df1 = Explorer.DataFrame.from_columns(x: [1, 2, 3], y: ["a", "b", "c"])
+      iex> df2 = Explorer.DataFrame.from_columns(x: [4.2, 5.3, 6.4], y: ["d", "e", "f"])
+      iex> Explorer.DataFrame.concat_rows([df1, df2])
+      #Explorer.DataFrame<
+        [rows: 6, columns: 2]
+        x float [1.0, 2.0, 3.0, 4.2, 5.3, "..."]
+        y string ["a", "b", "c", "d", "e", "..."]
+      >
   """
-  def concat_rows([%DataFrame{} = h | t] = dfs) do
-    key = Map.new(Enum.zip(names(h), dtypes(h)))
+  def concat_rows([%DataFrame{} | _t] = dfs) do
```

Comment:
I think we can simplify it a bit more. The trick to realize here is that, because all dataframes must have the same name and the compatible types, you can start with the initial shape and only do a single pass to compute the final shape. Something like:

    def compute_changed_types_concat_rows([head | tail]) do
      types = Map.new(Enum.zip(names(head), dtypes(head)))

      Enum.reduce(tail, %{}, fn df, changed_types ->
        if num_columns(df) != map_size(types) do
          raise "columns do not match"
        end

        Enum.reduce(Enum.zip(names(df), dtypes(df)), changed_types, fn {name, type}, changed_types ->
          cond do
            not Map.has_key?(changed_types, name) -> raise "columns do not match"
            types_are_equal?(...) -> changed_types
            types_are_compatible?(...) -> Map.put(changed_types, name, :float)
            true -> raise "incompatible types"
        end)
      end)